### PR TITLE
Implement source listing from debuginfo

### PR DESCRIPTION
**Pull request by Foxboron**
This pull-request does  a few things.

We implement a BuilID call in the RPC so the client can figure out which binary it's working on.

We implement support for fetching file listing through `debuginfod-find` if they are not present locally.

Then lastly we try to remove a bit of the duplicated code around dealing with `debuginfod-find`.

I'm not sure how much testing you want around this, or if new package belongs under `pkg/proc/debuginfod`. Feel free to nitpick and I'll try fix up all the issues around this patch :)

### CODE REVIEW

File: pkg/terminal/command.go [hunk 1]
```diff
@@ -2676,7 +2677,15 @@ func printfile(t *Term, filename string, line int, showArrow bool) error {
 		arrowLine = line
 	}
 
-	file, err := os.Open(t.substitutePath(filename))
+	var file *os.File
+	path := t.substitutePath(filename)
+	if _, err := os.Stat(path); errors.Is(err, os.ErrNotExist) {
```
Comment (aarzilli) [inline pkg/terminal/command.go:2682]: There's people around using old versions of go, let's just use `os.IsNotExist` here.

File: pkg/terminal/command.go [hunk 2]
```diff
@@ -2676,7 +2677,15 @@ func printfile(t *Term, filename string, line int, showArrow bool) error {
...
+		path, err = debuginfod.GetSource(t.client.BuildID(), filename)
```
Comment (aarzilli) [inline pkg/terminal/command.go:2683]: Only overwrite path if there is no error, if there is an error let the code continue as normal. Should this use a cache or is debuginfod-find supposed to be fast?

File: Documentation/cli/starlark.md [hunk 3]
```diff
@@ -20,6 +20,7 @@ Function | API Call
 amend_breakpoint(Breakpoint) | Equivalent to API call [AmendBreakpoint](https://godoc.org/github.com/go-delve/delve/service/rpc2#RPCServer.AmendBreakpoint)
 ancestors(GoroutineID, NumAncestors, Depth) | Equivalent to API call [Ancestors](https://godoc.org/github.com/go-delve/delve/service/rpc2#RPCServer.Ancestors)
 attached_to_existing_process() | Equivalent to API call [AttachedToExistingProcess](https://godoc.org/github.com/go-delve/delve/service/rpc2#RPCServer.AttachedToExistingProcess)
+build_i_d() | Equivalent to API call [BuildID](https://godoc.org/github.com/go-delve/delve/service/rpc2#RPCServer.BuildID)
```
[the remark this answers is not in the document]
Comment (Foxboron) [reply Documentation/cli/starlark.md:23]: Should I try improve it in this PR or?

File: pkg/terminal/command.go (hunk 1 above)
Comment (Foxboron) [reply pkg/terminal/command.go:2682]: go.mod explicitly says `go 1.16` so people on older versions can't use this regardless. I think it should be fine?

https://github.com/go-delve/delve/blob/master/go.mod#L3

File: Documentation/cli/starlark.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Foxboron) [reply Documentation/cli/starlark.md:23]: I could also do `BuildId` and omit the requires patch? It doesn't have to be `BuildID` even though it feels a bit more natural.

[the remark this answers is not in the document]
Comment (aarzilli) [reply Documentation/cli/starlark.md:23]: BuildId wouldn't follow our naming convention, doing the fix in this PR is ok, whatever you prefer.

File: pkg/terminal/command.go (hunk 2 above)
Comment (Foxboron) [reply pkg/terminal/command.go:2683]: We overwrite path if the path we get isn't available. `debuginfod-find` is fast enough and has it's own cache.

File: pkg/terminal/command.go (hunk 1 above)
Comment (aarzilli) [reply pkg/terminal/command.go:2682]: The go version in go.mod is not the minimum supported version of Go, delve does build on versions older than 1.16. It's true that we haven't bothered to keep track of how low we can get but errors.Is vs os.IsNotExist is too small of a difference to compromise.

Comment (Foxboron) [reply pkg/terminal/command.go:2682]: Ack, I'll change it :)

File: pkg/terminal/command.go (hunk 2 above)
Comment (aarzilli) [reply pkg/terminal/command.go:2683]: What I mean is, something like this:

```
path2, err := debuginfod.GetSource...
if err == nil {
    path = path2
}
```

so that the error message doesn't talk about debuginfod just because the source file can't be found and debuginfod-find can't be used.

Comment (Foxboron) [reply pkg/terminal/command.go:2683]: Ah yes. We can do something like that so the errors are consistent.

File: Documentation/cli/starlark.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (Foxboron) [reply Documentation/cli/starlark.md]: I looked at it and deemed it a fairly non-trivial change frankly. It seems like the current code doesn't deal with `CreateEBPFTracepoint` and has workaround for this. I applied the same workaround for `BuildID` and hope that is good enough.

https://github.com/go-delve/delve/blob/master/_scripts/gen-starlark-bindings.go#L114

File: pkg/proc/bininfo.go [hunk 4]
```diff
@@ -1393,8 +1355,37 @@ func (bi *BinaryInfo) loadSymbolName(image *Image, file *elf.File, wg *sync.Wait
...
+		bi.logger.Error("invalid build-id signature")
+		return
+	}
+
+	descBinary := make([]byte, bh.Descsz)
+	if err := binary.Read(br, binary.LittleEndian, descBinary); err != nil {
+		bi.logger.Errorf("can't read build-id desc: %w", err)
+		return
+	}
+	desc := hex.EncodeToString(descBinary)
```
Comment (aarzilli) [inline pkg/proc/bininfo.go:1387]: bi.BuildID = hex.EncodeToString(descBinary)

**Review by aarzilli — COMMENTED** [review]
See build failure (bi.logger.Errorf doesn't support the %w verb)

**Comment by Foxboron** [comment]
Woops, sorry. I should probably not be lazy with the tests :)

Amended the commit with the fixes.

**Review by aarzilli — APPROVED** [review]
LGTM

File: service/client.go [hunk 5]
```diff
@@ -12,6 +12,9 @@ type Client interface {
 	// Returns the pid of the process we are debugging.
 	ProcessPid() int
 
+	// Returns the BuildID of the process' exectuable we are debugging.
```
Comment (stefanvanburen) [inline service/client.go:15]:
```suggestion
	// Returns the BuildID of the process' executable we are debugging.
```

File: pkg/proc/bininfo.go [hunk 6]
```diff
@@ -52,6 +52,9 @@ type BinaryInfo struct {
 
 	debugInfoDirectories []string
 
+	// BuilID of this binary.
```
Comment (stefanvanburen) [inline pkg/proc/bininfo.go:55]:
```suggestion
	// BuildID of this binary.
```

😄 sorry for the drive-bys!

Comment (Foxboron) [reply pkg/proc/bininfo.go]: Thanks :)!

**Review by derekparker — APPROVED** [review]
LGTM

**Comment by TobixDS** [comment]
Hi,
is it correct that the required ELF section is 
".note.gnu.build-id"?

If inspect any go executable (e.g. go itself - version go1.16.13 linux/amd64) I find only a section without a dash:

```
readelf -n /usr/local/go/bin/go

Displaying notes found in: .note.go.buildid
  Owner                Data size        Description
  Go                   0x00000053       GO BUILDID

```

Therefor delve 1.8.1 complains about the missing build id

```
dlv version
Delve Debugger
Version: 1.8.1
Build: $Id: d85f1f6b736db99a1c239c34fd4a081dcff08a3c $

dlv exec /usr/local/go/bin/go
2022-02-09T11:11:03Z error layer=debugger can't find build-id note on binary

dlv exec /go/bin/dlv         
2022-02-09T11:17:03Z error layer=debugger can't find build-id note on binary
```

**Comment by Foxboron** [comment]
It is correct. Debuggers don't understand `.note.go.buildid` and what you are looking at is the fact that not all Go binaries has a build-id embedded. I'm not sure if there is any good fixes for this.

**Comment by Foxboron** [comment]
Right, it's a feature of cgo and Go itself doesn't support it.

https://github.com/golang/go/issues/41004

https://github.com/kata-containers/kata-containers/issues/2587#issuecomment-915705151
